interp/notation.ml

```diff
@@ -675,10 +675,8 @@ let rec constr_of_glob allow_constant to_post post env sigma g = match DAst.get
               let sigma,cl = aux sigma a gcl in
               sigma, c :: cl
            | ToPostCheck r :: a, gc :: gcl ->
-              let () = match DAst.get gc with
-                | Glob_term.GRef (r', _) when GlobRef.equal r r' -> ()
-                | _ -> raise NotAValidPrimToken in
               let sigma,c = constr_of_glob true [||] [] env sigma gc in
+              if not (Constr.equal r c) then raise NotAValidPrimToken;
```

Comment:
This looks very fragile to me. Term equality is notoriously wobbly, e.g. this might fail because of universe instances and the like. Can't we simply allow an arbitrary term?

Reply:
Allowing an arbitrary term means we could call the printing function on the wrong type (imagine a parsing function that parses `option bool`, calling it on `option nat` would be wrong), ideally we should check for convertibility.

Reply:
Then I'd rather have a small DSL of patterns allowing nested applications of references and wildcards than arbitrary terms. This is really asking for trouble, we've so many issues with arbitrary term equality...

Reply:
I don't want to derail this PR by increasing it's scope (I'm happy to have either the current version or @ppedrot's suggestion, ideally whichever one we go with coming with a comment about pitfalls and future extension directions), but I wonder if the correct time to check this is actually via conversion when constructing the application, rather than here.  That is, permit an arbitrary term here, retype the application of the parsing function to the term (after checking that it's closed?), and then reduce.  This might slot in nicely with permitting overlapping number notations, such that if a given one either fails to typecheck or returns `None`, we fall back to the next one.  (I don't have a use case yet for this, though, so perhaps it's not worth the complexity.)

Reply:
The problem is that not failing early can be catastrophic for performance. Furthermore, we need the check to be fast, ideally O(1) and worst case O(n), otherwise printing is going to be laboriously slow in some corner cases.

Reply:
@ppedrot here is a solution without arbitrary term equality and failing as early as possible
@JasonGross sorry, I'm not sure I grasp your idea, would you have an illustrative example? (even a dummy one)

Reply:
The current code seems more reasonable to me.